### ISSUE TITLE
Operator "+" not supported for types "int | Tuple[int, ...]" and "int"

### ISSUE THREAD

**Issue by kumkee** [microsoft/pyright#4170]
**Describe the bug**
If my function returns int or tuple, and I am sure it will give me an int, pyright still complains about adding two different types.
One common example is the overridden `__getitem__` function. See an example below

**To Reproduce**
```
"""Reproduce a pyright bug."""
import typing as t


class MyClass:
    """Dummy class to reproduce a bug."""

    def __init__(self, content=tuple(range(9))):
        """init."""
        self.content = content

    def __getitem__(
        self, key: t.Union[int, slice]
    ) -> t.Union[int, t.Tuple[int, ...]]:
        """My __getitem__."""
        return self.content[key]


i: int = 4
obj = MyClass()
print(i + obj[i])
```
Using typing.Union and so is for python backward compatibility.
Pyright complains:
`Operator "+" not supported for types "Literal[4]" and "int | Tuple[int, ...]"`

**Expected behavior**
Because the key is an int, I would expect pyright is clever enough to tell the result from `__getitem__` is a single item.

**Screenshots or Code**
In the longer code I am working on, the `i` is unknown. It gave me:
`Operator "+" not supported for types "int | Tuple[int, ...]" and "int"`

**VS Code extension or command-line**
I am using Neovim and pyright 1.1.279 as a plugin. It likely runs in the background as a command-line app.

**Additional context**
N.A.

**Comment by Ovsyanka83** [comment]
Pyright is correct here. If you tell pyright that your method returns int or tuple, then pyright will listen to you and enforce it. 

Allow me to demonstrate how your code can break:
```python
"""Reproduce a pyright bug."""
import typing as t


class MyClass:
    """Dummy class to reproduce a bug."""

    def __init__(self, content=tuple(range(9))):
        """init."""
        self.content = content

    def do_something_sneaky(self):
        self.content += ((1,),)

    def __getitem__(self, key: t.Union[int, slice]) -> t.Union[int, t.Tuple[int, ...]]:
        """My __getitem__."""
        return self.content[key]


i = 4
obj = MyClass()
print(i + obj[i])

obj.do_something_sneaky()

i = 9
print(i + obj[i])
```

And now you get the TypeError pyright warned you about. And this code is perfectly valid.

If you believe that pyright should know better and use the values in your code, then think about how pyright would handle the following cases:
1) the use of eval()
2) the use of variable input (random, input(), requests.get, file reading)
3) the use of untyped code within typed code)

So pyright is doing the same thing any other typechecker (mypy, pyre) would do: it's mostly analyzing types and trusting you with writing your types correctly. I suggest the use of typing.TypeGuard or typing.overload if you want more precision.

**Comment by kumkee** [comment]
TypeGuard would require python 3.10, but type overload should work.
Thanks for the prompt and helpful answer.

**Comment by Ovsyanka83** [comment]
@kumkee https://pypi.org/project/typing-extensions/ is here to help you! It is a backport that includes all typing features from the most modern python versions, and pyright fully supports it.

**Comment by kumkee** [comment]
> @kumkee https://pypi.org/project/typing-extensions/ is here to help you! It is a backport that includes all typing features from the most modern python versions, and pyright fully supports it.

Thanks for letting me but my runtime environment doesn't allow anything beyond the std library. 

I've tried typing overload, it works for the toy class above, but not for the one I am working on. Here is my class
```
@dataclass
class Stack:
    """Tuple to represent a stack of cards with repeatable parts."""

    core: t.Tuple[int, ...] = ()
    repeatitions: t.Counter[t.Tuple[int, int]] = field(default_factory=Counter)

    def __repr__(self) -> str:
        """Override __repr__."""
        # Fancier repr. Not essential.
        reprn = str(self.core)
        for pair, num in self.repeatitions.items():
            reprn = insert_repetition(reprn, pair, num)
        return reprn

    @t.overload
    def __getitem__(self, key: int) -> int:
        """Take integer input."""

    @t.overload
    def __getitem__(self, key: slice) -> t.Tuple[int, ...]:
        """Take slice input."""

    def __getitem__(self, key):
        """Override getitem."""
        return self.core[key]

    def __add__(self, other: t.Tuple[int, ...]):
        """Override add."""
        ret = copy(self)
        ret.core += other
        return ret

    def __bool__(self) -> bool:
        """Override bool."""
        return self.core != ()

    def __len__(self) -> int:
        """Override len."""
        return len(self.core)
```
At the actual implementation, I got this error message
```
Overloaded implementation is not consistent with signature of overload 2
Function return type "Tuple[int, ...]" is incompatible wi Pyright (reportGeneralTypeIssues) [39, 9]
```

**Comment by erictraut** [comment]
You have not provided type annotations for the `__getitem__` implementation. It should accept a combination of all possible parameters and types allowed by the overloads.

```python
    def __getitem__(self, key: int | slice) -> int | tuple[int, ...]: ...
```

**Comment by kumkee** [comment]
Oop. Thank you @erictraut
